arlo_server/rounds.py

```diff
@@ -0,0 +1,154 @@
+from flask import jsonify, request
+from jsonschema import validate
+from werkzeug.exceptions import BadRequest, Conflict
+import uuid, datetime
+
+from arlo_server import app, db
+from arlo_server.models import (
+    Round,
+    RoundContest,
+    Election,
+    SampledBallot,
+    SampledBallotDraw,
+)
+from arlo_server.auth import with_election_access, UserType
+from arlo_server.sample_sizes import sample_size_options
+from util.isoformat import isoformat
+from audits import sampler
+
+
+CREATE_ROUND_REQUEST_SCHEMA = {
+    "type": "object",
+    "properties": {
+        "roundNum": {"type": "integer", "minimum": 1,},
+        "sampleSize": {"type": "integer", "minimum": 1,},
+    },
+    "additionalProperties": False,
+    "required": ["roundNum"],
+}
+
+
+def serialize_round(r: Round) -> dict:
+    return {
+        "id": r.id,
+        "roundNum": r.round_num,
+        "startedAt": isoformat(r.started_at),
+        "endedAt": isoformat(r.ended_at),
+    }
+
+
+# Raises if invalid
+def validate_round(r: dict, election: Election):
+    validate(r, CREATE_ROUND_REQUEST_SCHEMA)
+
+    rounds = sorted(election.rounds, key=lambda r: r.round_num, reverse=True)
+    current_round = next(iter(rounds), None)
+    next_round_num = current_round.round_num + 1 if current_round else 1
+    if r["roundNum"] != next_round_num:
+        raise BadRequest(f"The next round should be round number {next_round_num}")
+
+    if current_round and not current_round.ended_at:
+        raise Conflict(f"The current round is not complete")
+
+    if r["roundNum"] == 1 and "sampleSize" not in r:
+        raise BadRequest(f"Sample size is required for round 1")
+
+
+def sample_ballots(election: Election, round: Round, sample_size: int):
+    # For now, we only support one targeted contest
+    contest = next(c for c in election.contests if c.is_targeted)
+
+    # Compute the total number of ballot samples in all rounds leading up to
+    # this one. Note that this corresponds to the number of SampledBallotDraws,
+    # not SampledBallots.
+    previous_rounds = (
+        db.session.query(RoundContest)
+        .join(Round)
+        .filter(Round.election_id == election.id, Round.id != round.id)
+        .all()
+    )
+    num_sampled = sum(r.sample_size for r in previous_rounds)
```

Comment:
@benadida @umbernhard please check this. The single-jurisdiction flow computed this by actually counting `SampledBallotDraw`s in the db, but I thought this would be a bit simpler of a db query. I want to be sure they are equivalent.

Reply:
They should be the same, but given that the instructions we give audit boards are taken from `SampledBallotDraw`, I have a small preference to use that as the reference for how many were sampled. Small preference, your call.

Reply:
It ended up being simpler that way haha! ba5fbe7